template/source/javascripts/modules/in-page-navigation.js

```diff
@@ -0,0 +1,93 @@
+(function($, Modules) {
+  'use strict';
+
+  Modules.InPageNavigation = function InPageNavigation() {
+    var $tocPane;
+    var $contentPane;
+    var $tocItems;
+
+    var animationFrameRequested = false;
+
+    this.start = function start($element) {
+      if (!Modernizr.history) {
+        return;
+      }
+
+      $tocPane = $element.find('.app-pane__toc');
+      $contentPane = $element.find('.app-pane__content');
+      $tocItems = $tocPane.find('a');
+
+      $contentPane.on('scroll', function () {
+        withAnimationFrame(handleScrollEvent);
+      });
+
+      // Popstate is triggered when using the back button to navigate 'within'
+      // the page, i.e. changing the anchor part of the URL.
+      $(window).on('popstate', function (event) {
+        restoreScrollPosition(event.originalEvent.state);
+      });
+
+      // Restore state when e.g. using the back button to return to this page
+      restoreScrollPosition(history.state);
+    };
+
+    function restoreScrollPosition(state) {
+      if (state && state.scrollTop) {
+        $contentPane.scrollTop(history.state.scrollTop);
+      }
+    }
+
+    function handleScrollEvent() {
+      var activeTocItem = tocItemForFirstElementInView();
+
+      storeCurrentPositionInHistoryApi(activeTocItem);
+      highlightActiveItemInToc(activeTocItem);
+    }
+
+    function storeCurrentPositionInHistoryApi(activeTocItem) {
+      history.replaceState(
+        { scrollTop: $contentPane.scrollTop() },
+        "",
+        activeTocItem.attr('href')
+      );
+    }
+
+    function highlightActiveItemInToc(activeTocItem) {
+      $tocItems.removeClass('toc-link--in-view');
+
+      if (activeTocItem) {
+        activeTocItem.addClass('toc-link--in-view');
+      }
+    }
+
+    function tocItemForFirstElementInView() {
+      var target = null;
+
+      $($tocItems.get().reverse()).each(function checkIfInView(index) {
+        if (target) {
+          return;
+        }
+
+        var $this = $(this);
+        var $heading = $contentPane.find($this.attr('href'));
+
+        if ($heading.position().top <= 0) {
+          target = $this;
+        }
+      });
+
+      return target;
+    }
+
+    function withAnimationFrame(callback) {
```

Comment:
I wonder if it's a better idea to use an off-the-shelf debounce function rather than rolling our own – in the past I've included [lodash.debounce](https://www.npmjs.com/package/lodash.debounce). That's a node module, but you can generate a lodash build with just the debounce function via `lodash -d include=debounce exports=global` (after `npm install -g lodash-cli`).

Ironically that doesn't actually use `requestAnimationFrame`, so this implementation is perhaps a touch more performant – I'll leave it up to you.